modules/imgproc/src/stackblur.cpp

```diff
@@ -0,0 +1,1261 @@
+            v_store(sumOut + VEC_LINE_32 + k, v_sumOut1);
+
+            v_sumIn0 -= v_reinterpret_as_s32(v_ss0);
+            v_sumIn1 -= v_reinterpret_as_s32(v_ss1);
+
+            v_store(sumIn + k, v_sumIn0);
+            v_store(sumIn + VEC_LINE_32 + k, v_sumIn1);
+        }
+    }
+    return k;
+}
+#endif
+
+template<typename T, typename TBuf>
+class ParallelStackBlurColumn:
+        public ParallelLoopBody
+{
+public:
+    ParallelStackBlurColumn (const Mat & _src, Mat &_dst, int _radius):src(_src), dst(_dst) ,radius(_radius)
+    {
+        CN = src.channels();
+        widthElem = CN * src.cols;
+        height = src.rows;
+        hm = src.rows - 1;
+        mulVal = 1.0f / ((radius + 1)*(radius + 1));
+        if (radius <= STACKBLUR_MAX_RADIUS)
+        {
+            shrValTab = stackblurShr[radius];
+            mulValTab = stackblurMul[radius];
+        }
+        else
+        {
+            shrValTab = 0;
+            mulValTab = 0;
+        }
+    }
+
+    ~ParallelStackBlurColumn() {}
+
+    ParallelStackBlurColumn& operator=(const ParallelStackBlurColumn &) { return *this; }
+
+    virtual void operator ()(const Range& range) const CV_OVERRIDE
+    {
+        if (radius == 0)
+            return;
+
+        const int kernelSize = 2 * radius + 1;
+        int widthImg = std::min(range.end, src.cols * CN);
+        int widthLen = widthImg - range.start;
+
+        size_t bufSize = 3 * widthLen * sizeof(TBuf) + kernelSize * widthLen * sizeof(T);
+
+        AutoBuffer<uchar> _buf(bufSize + 16);
+        uchar* bufptr = alignPtr(_buf.data(), 16);
+
+        TBuf* sum = (TBuf *)bufptr;
+        TBuf* sumIn = sum + widthLen;
+        TBuf* sumOut = sumIn + widthLen;
+        T* stack = (T* )(sumOut + widthLen);
+
+        memset(bufptr, 0, bufSize);
+
+        const T* srcPtr =dst.ptr<T>() + range.start;
+
+        for (int i = 0; i <= radius; i++)
+        {
+            for (int k = 0; k < widthLen; k++)
+            {
+                stack[i * widthLen + k] = *(srcPtr + k);
+                sum[k] += *(srcPtr + k) * (i + 1);
+                sumOut[k] += *(srcPtr + k);
+            }
+        }
+
+        for (int i = 1; i <= radius; i++)
+        {
+            if (i <= hm) srcPtr += widthElem;
+            for (int k = 0; k < widthLen; k++)
+            {
+                T tmp = *(srcPtr + k);
+                stack[(i + radius) * widthLen + k] = tmp;
+                sum[k] += tmp * (radius - i + 1);
+                sumIn[k] += tmp;
+            }
+        }
+
+        int sp = radius;
+        int yp = radius;
+
+        if (yp > hm) yp = hm;
+
+        T* dstPtr = dst.ptr<T>() + range.start;
+        srcPtr = dst.ptr<T>(yp) + range.start;
+        int stackStart = 0;
+
+        for(int i = 0; i < height; i++)
+        {
+            stackStart = sp + kernelSize - radius;
+            if (stackStart >= kernelSize) stackStart -= kernelSize;
+
+            int sp1 = sp + 1;
+            sp1 &= -(sp1 < kernelSize);
```

Comment:
Are you sure that compilers optimize this properly?

Reply:
The no optimized code is following:
```
int sp1 = sp + 1;
if (sp1 >= kernelSize)
    sp1 = 0;
```

Reply:
Yes, and we should not obfuscate code for compilers.
More operations may block optimizations (because we can't pass assumptions into optimization transformations).

See here: https://godbolt.org/z/G3jdx5v8a